### PR TITLE
feat!: support multiple virtual networks per module instance

### LINKED ISSUE #97: Feature Request: Allow for more than one vnet per subscription
- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Description

**Is your feature request related to an issue?**

No open issue but we cannot use the module to build more than one vnet per subscription

### Describe the solution you'd like

We would like the ability to add more than one vnet per subscription. The use case for this is when you have one subscription but using resources in multiple regions. For example:

Subscription: "AVD"
Vnet1: "vnet-avd-uksouth"
Vnet2: "vnet-avd-ukwest"

At present the yaml /module only supports one vnet (example)

```
display_name: avd
name: avd
workload: Production
location: uksouth
management_group_id: lz-avd
vnet_rg_name: rg-vnet-avd
vnet_name: vnet-avd-uksouth
vnet_address_space:
  - "10.0.3.0/24"
  - "192.168.3.0/24"
```

It looks like vnets could use a terraform map the same way that role assigments does. 

As a workaround we will be trying to use the [Example 5 Use with existing subscriptions ](https://github.com/Azure/terraform-azurerm-lz-vending/wiki/Example-5-Use-with-existing-subscriptions ) to add more vnets or native terraform. 

### Additional context
Module is brilliant, keep up the good work :)

### LINKED ISSUE #99: Feature Request: Support for resource group and vnet tags
- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Description

**Is your feature request related to an issue?**

We have a policy which enforces tags, at present this module doesn't support tags on resource groups or vnets. 

### Describe the solution you'd like

Add support for tags on resource groups and vnets. 

 We have a policy which enforces tags which means we have to delete the assignment / add an exception to allow resource group or vnets to be built. 

### Additional context
I did see that #62 added support to use tags with subscriptions.

### LINKED ISSUE #101: Feature Request: Add vnet_id as an output
- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Description

**Is your feature request related to an issue?**
No

We would like to use the vnet_id of any vnets that are generated from this vending module without having to use a data source. 

### Describe the solution you'd like

[outputs.tf](https://github.com/Azure/terraform-azurerm-lz-vending/blob/main/outputs.tf) to include the vnet_id of any vnets it creates. At present it only outputs the subscription_id's. It looks like vnet_id is in the [output.tf](https://github.com/Azure/terraform-azurerm-lz-vending/blob/main/modules/virtualnetwork/outputs.tf) from the virtualnetwork submodule 

### Additional context
This will allow us to use this vending module to create a 'connectivity' subscription and then pass the vnet_id into any other subscriptions for vnet peering them back to the connectivity.

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
# Overview/Summary

PR to add multiple vnet support. Also adding requested features such as tagging for resources and resource groups as well as virtual network resource id outputs.

## This PR fixes/adds/changes/removes

1. fixes #97 
2. fixes #101 
3. fixes #99 

### Breaking Changes

1. Updated variable schema to support multiple virtual networks

## Testing Evidence

* Tests updated to support new var inputs
* Additional tests for new variables, e.g. custom naming

Deployment tests run outside GH actions due to #61

image

image

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/terraform-azurerm-lz-vending/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/terraform-azurerm-lz-vending/issues), for tracking and closure.
- [x] Run and `make fmt` & `make docs` to format your code and update documentation.
- [x] Created unit and deployment tests and provided evidence.
- [x] Updated relevant and associated documentation.
- [ ] (Optionally) Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located in the [Enterprise-Scale repo](https://github.com/Azure/Enterprise-Scale) in the directory: `/docs/wiki/whats-new.md`)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,8 +11,10 @@ The module is designed to be instantiated many times, once for each desired land
 This is currently split logically into the following capabilities:
 
 - Subscription creation and management group placement
-- Hub & spoke networking
-- Virtual WAN networking
+- Networking - deploy multiple vnets with:
+  - Hub & spoke connectivity
+  - vWAN connectivity
+  - Mesh peering (peering between spokes)
```
Comment (krowlandson) [inline README.md:17]: Question: can you include hubs as part of the mesh? Should we link to the vnet peering limits?

File: variables.virtualnetwork.tf [hunk 2]
```diff
@@ -1,142 +1,113 @@
...
-DESCRIPTION
-  default     = true
-}
+### Other peerings
 
-variable "virtual_network_vwan_associated_routetable_resource_id" {
-  type        = string
-  description = <<DESCRIPTION
-The resource ID of the virtual network route table to use for the virtual network.
+TODO
```
Comment (krowlandson) [inline variables.virtualnetwork.tf:88]: I assume you are skipping this on the PR on purpose?

**Review by krowlandson — DISMISSED** [review]
Looks great... I'm approving on the assumption that deployment works based on the automated tests.

A couple of comments, but nothing that would require an update unless you want to include now.

Only thing I was wondering is what the upgrade experience is like? Should we provide guidance for users who are upgrading from a previous version? Does this have any considerations give the input changes?

**Comment by matt-FFFFFF** [comment]
Thanks @krowlandson 

The deployment tests do indeed pass.

I will add a guide to the on the upgrade path and also resolve some of the other comments.

File: variables.virtualnetwork.tf (hunk 2 above)
Comment (matt-FFFFFF) [reply variables.virtualnetwork.tf:88]: Removed - thanks

**Review by krowlandson — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**